keps/0018-controller-overhaul.md

```diff
@@ -0,0 +1,121 @@
+---
+kep-number: 18
+title: Controller redesign
+authors:
+  - "@alenkacz"
+owners:
+  - "@alenkacz"
+creation-date: 2019-09-16
+last-updated: 2019-09-16
+status: provisional
+---
+
+# Controller redesign
+
+## Table of Contents
+
+* [Controller redesign](#controller-redesign)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [Instance CRD changes](#instance-crd-changes)
+         * [Admission webhook](#admission-webhook)
+         * [Implementation details](#implementation-detailsnotesconstraints)
+         * [Risks and mitigations](#risks-and-mitigations)
+      * [Graduation Criteria](#graduation-criteria)
+
+## Summary
+
+The way KUDO controllers are currently implemented has several flaws:
+1. When KUDO manager is down (or restarted) we might miss an update on a CRD meaning we won’t execute the plan that was supposed to run or we execute an incorrect one ([issue](https://github.com/kudobuilder/kudo/issues/422))
+2. Multiple plan can be seemingly in progress leading to misleading status communicated ([issue](https://github.com/kudobuilder/kudo/issues/628))
+3. No way to ensure atomicity on plan execution (given the current CRD design where information is spread across several CRDs)
+4. Very low test coverage and overall confidence in the code
```

Comment:
Are there any linters/metrics we can put together? Not as something we have to _live_ by but just to give us a general sense in data of whether or not we are improving. Maybe a separate KEP, but I don't think we need to institutionalize anything, just collect data as we go even if we're running it manually.

Reply:
you mean around coverage? Overall I don't like measuring code coverage, getting that number higher not always improves things. Right now there are 0 unit tests, my goal was to go above that.

But I know that the claim I made here sounds very generic and it would be good to have something objective that would mean it's good enough... I am open to suggestions here :)

Reply:
Nah, not coverage. I'm thinking simple gut checks we can run, like running gocyclo and a few other stats. Basically, we should have a general sense that are we improving those main functions by driving different sources complexity down. Totally agree that code coverage on its own doesn't really tell us anything.